pkg/scheduler/framework/plugins/dynamicresources/structured/namedresources/namedresourcesmodel.go

```diff
@@ -0,0 +1,160 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package namedresources
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"slices"
+
+	resourceapi "k8s.io/api/resource/v1alpha2"
+	"k8s.io/apiserver/pkg/cel/environment"
+	resourceinternal "k8s.io/kubernetes/pkg/apis/resource"
+	"k8s.io/kubernetes/pkg/apis/resource/structured/namedresources/cel"
+	resourceinternalversioned "k8s.io/kubernetes/pkg/apis/resource/v1alpha2"
+)
+
+type Model struct {
+	instances []instanceAllocation
+}
+
+type instanceAllocation struct {
+	allocated bool
+	instance  *resourceinternal.NamedResourcesInstance
+}
+
+// AddResources must be called first to create entries for all existing
+// resource instances. The resources parameter may be nil.
+func AddResources(m *Model, resources *resourceapi.NamedResourcesResources) {
+	if resources == nil {
+		return
+	}
+
+	// Evaluating the CEL expressions needs the internal representation of the
+	// instance attributes. In practice this is a shallow copy of the instance
+	// slice and thus cheap.
+	var internalResources resourceinternal.NamedResourcesResources
```

Comment:
Note to self: it feels wrong that code working with the external API needs to call the "internal" code with the "internal" types.

It's probably better to move `k8s.io/kubernetes/pkg/apis/resource/structured/namedresources/cel` into `k8s.io/dynamic-resource-allocation/structured/namedresources/cel` and change its API so that it can work with both internal and external types, i.e. access attributes through an interface instead of some specific type.

Once it is in the staging repo, vendors can use it in unit tests to verify the CEL code that they generate in combination with different attributes.

Reply:
Resolved...